### PR TITLE
Add WIP support for returning top tokens

### LINKED ISSUE #604: Return most probable tokens + logprobs
### Feature request

It would be great if the API could return a list of most probable tokens (along with their logprobs) for each step. This could be useful for many downstream tasks that require sampling of these. 

### Motivation

Many LLM-based applications benefit not just from having the most probable token, but also a list of output probabilities.

* Some classification tasks can be done by selecting only the most probable token, but only when this is actually one of the possible labels. Having the output logits, makes it possible to mitigate this and just select the token that's actually one of the options (having logit bias available will also help with this, but that's another feature)
  * `lm-evaluation-harness` relies on having the output logits for its multiple-choice benchmarks. Having this feature will allow using this server to speed up benchmarking tasks.
* In many classification use-cases you want the probability of the answer. Without `logit_bias`, you could use `top_k`, hope your prompting is sufficient such that the first `k` tokens are all your other classes and that the `softmax` is only taken over these tokens. It is, however, impossible to know for sure.
* For the same reason as the previous point, binary ranking is unreliable. Multi-class ranking is nigh impossible (except for using high `best-of` and `temperature` to hope you get all label's probabilities). 

### Your contribution

I'd be willing to contribute to this feature, nut I'd love some guidance on that. Some initial reading of the code makes me think this won't require big changes changes in the code, though it'll touch many different parts of it. For that reason I'd want to be sure of the preferred approach before wasting both my and your time.

I guess we could add this in (all variations of) [Model#generate_token](https://github.com/huggingface/text-generation-inference/blob/b7327205a6f2f2c6349e75b8ea484e1e2823075a/server/text_generation_server/models/flash_causal_lm.py#L768) and add the top n values from `next_token_logits` to this iterator:

https://github.com/huggingface/text-generation-inference/blob/b7327205a6f2f2c6349e75b8ea484e1e2823075a/server/text_generation_server/models/flash_causal_lm.py#L891-L902

And then also add it to the `Generation`?

https://github.com/huggingface/text-generation-inference/blob/b7327205a6f2f2c6349e75b8ea484e1e2823075a/server/text_generation_server/models/flash_causal_lm.py#L974-L982

Provided we make sure the input and output is passed on correctly everywhere else where it's touched, would this be the way to go?

I'm not a Rust programmer, but the code seems straightforward enough (famous last words?). Just add the parameter everywhere where the other params like `top_k` are also mentioned.

---

Regarding how to call this feature. OpenAI has an integer parameter called "logprobs" that does just this. However, this won't work since TGI already has another parameter with this name.  Perhaps `top_n_logprobs`, `top_k_logprobs`, `top_n_tokens`, `top_k_tokens`?

### PR DESCRIPTION
**Pull request by Vinno97**
# What does this PR do?

This PR add support for return the log probabilities of the top N tokens (besides the chosen model).

The token selection code is loaned from the [IBM's fork](https://github.com/IBM/text-generation-inference/) (much thanks @njhill) and I merged into the current `main` branch. I chose to name the input parameter `top_n_tokens` and it can be any (positive) integer. If it's set, a `top_tokens` field in the result `details` will contain a LxN list of tokens where L is the sequence length and N is the amount of requested tokens.

Note that it is currently very much WIP.  

* It is currently only implemented for seq-to-seq models (though it's modular and should be easy to implement for other models as well). 
* I haven't implemented this functionality for the prefill. **Do we want to?**
* I sort the tokens based on their `logprobs`, because IBM's implementation had the functionality. However, it may cause unnecessary overhead. 
* IBM's code doesn't just return the "top n" tokens. It actually (explicitly) returns more tokens when there are tokens with equal probabilities. I am not sure if this is desired.
* I have not yet implemented input validation of the `top_n_tokens` field.
* I have also done no benchmarking to test the impact on performance. 
  * I haven't given the benchmarking tool the capability to benchmark this feature.

Fixes #604 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. 
Specifically @OlivierDehaene and @njhill

### CODE REVIEW

**Comment by spew** [comment]
Could this be used to "shorten" a generation (i.e. only use the first N tokens)? If so, how would one use the logprobs to calculate a threshold for throwing out all or a portion of a generation?

**Comment by Vinno97** [comment]
This PR only adds the ability to not just return the sampled token, but also the most likely alternative tokens. This can be useful mostly when building applications that require the logprobs of multiple tokens, like ranking and classification

 What you are referring to, sounds like a custom stopping criterium.

**Comment by OlivierDehaene** [comment]
This is very cool. Thanks for the work!
Are you ok with me maye merging this to a dev branch to add my modifications directly before merging it to the main branch? 
Cheers!

**Comment by Vinno97** [comment]
It currently only works (properly) for `FlashCausalLM` models. Perhaps it'd be good to wait until I implement it everywhere

**Comment by Vinno97** [comment]
@OlivierDehaene I think it's mergable now!

Haven't edited the Python Client yet, but all server functionalities should work now. For some reason, my latest commit (95d0fba7de062840dde979c043581085d4a7ac9d) doesn't show up yet in this PR, however.

**Comment by OlivierDehaene** [comment]
@Vinno97, I will be off in August but @Narsil will be available to review the PR.
For your info, we updated the license of this repo to HFOIL. You can review the reasons behind this license change [here](https://github.com/huggingface/text-generation-inference/issues/726).

**Comment by Vinno97** [comment]
Thank you for notifying me on the licencing change. Luckily, as far as I understand it, this has no effect on our operations for now

**Comment by Vinno97** [comment]
Hi @Narsil. Could you review this PR? It touches a lot of different files though most changes are only about passing parameters and results have to be passed through every communication phase. The token logic is contained in [`tokens.py#batch_top_tokens` ](https://github.com/huggingface/text-generation-inference/blob/e30f4f61e7958ba7a29924705cb6ffcc9766ec3d/server/text_generation_server/utils/tokens.py#L339). Additionally, I added `top_n_token` support separately to `CausalLM`, `FlashCausalLM` and `Seq2SeqLM` .

**Comment by Narsil** [comment]
Thanks for the PR. 

I'm pretty stretched atm and this PR looks like it requires a lot of attention.
I intend to look at it early next week. Don't hesitate to ping if I didn't.

**Comment by Vinno97** [comment]
Understandable. 
In the meantime, we started running it in our testing/experimentation enviroment and a colleague of mine is integrating it with a private fork of lm-eval. If we find any issues there, I'll of course update the PR.

**Comment by Vinno97** [comment]
Just fixed a small oversight in the batch concatenation for flash attention models

**Comment by Vinno97** [comment]
> I intend to look at it early next week. Don't hesitate to ping if I didn't.

@Narsil ping 🙂

Since it's my PR, I obviously can't review myself. But let me know if I can be of any help.

**Comment by Narsil** [comment]
Ok I looked at the code. On the surface the code looks really nice ! 

I need to invest a bit of time investigating the performance changes on various models/platforms.
At least making sure there's no huge regression.

**Comment by Narsil** [comment]
Are you OK if I do the rebase to be up-to-date too ?

**Comment by Narsil** [comment]
Ok here are the results from the bench: 

https://github.com/Narsil/tmp_bench/blob/main/README.md
Performance is affected mostly for smaller models without TP.
It kind of makes sense since the less compute on the model, the more compute actually happens during logits processing. I'm still surprised by the 25%.

At least there doesn't seem to be any degradation without using `top_p_tokens`, so it should be good for most users (and users can deactivate the option)

Overall I find everything proposed in the PR acceptable.
Default to top_k=5 tokens seems reasonable (bench used 10).

**Comment by Narsil** [comment]
Here is my rebased version: https://github.com/huggingface/text-generation-inference/pull/868

**Comment by Vinno97** [comment]
Closing since @Narsil's merge via https://github.com/huggingface/text-generation-inference/pull/868
